util/types/time.go

```diff
@@ -1542,3 +1538,145 @@ func IsDateFormat(format string) bool {
 func ParseTimeFromInt64(num int64) (Time, error) {
 	return parseDateTimeFromNum(num)
 }
+
+// Format returns a textual representation of the time value formatted
+// according to layout.
+// See: http://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format
+func (t Time) Format(layout string) (string, error) {
+	var buf bytes.Buffer
+	inPatternMatch := false
+	for _, b := range layout {
+		if inPatternMatch {
+			if err := t.convertDateFormat(b, &buf); err != nil {
+				return "", errors.Trace(err)
+			}
+			inPatternMatch = false
+			continue
+		}
+
+		// it's not in pattern match now
+		if b == '%' {
+			inPatternMatch = true
+		} else {
+			buf.WriteRune(b)
+		}
+	}
+	return buf.String(), nil
+}
+
+var abbrevMonthName = []string{
+	"Jan", "Feb", "Mar", "Apr", "May", "Jun",
+	"Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
+}
+
+var abbrevWeekdayName = []string{
+	"Sun", "Mon", "Tue",
+	"Wed", "Thu", "Fri", "Sat",
+}
+
+func (t Time) convertDateFormat(b rune, buf *bytes.Buffer) error {
+	switch b {
+	case 'b':
+		m := t.Time.Month()
+		if m == 0 || m > 12 {
+			return errors.Trace(ErrInvalidTimeFormat)
+		}
+		buf.WriteString(abbrevMonthName[m-1])
+	case 'M':
+		m := t.Time.Month()
+		if m == 0 || m > 12 {
+			return errors.Trace(ErrInvalidTimeFormat)
+		}
+		buf.WriteString(MonthNames[m-1])
+	case 'm':
+		fmt.Fprintf(buf, "%02d", t.Time.Month())
+	case 'c':
+		fmt.Fprintf(buf, "%d", t.Time.Month())
+	case 'D':
+		fmt.Fprintf(buf, "%d%s", t.Time.Day(), abbrDayOfMonth(t.Time.Day()))
+	case 'd':
+		fmt.Fprintf(buf, "%02d", t.Time.Day())
+	case 'e':
+		fmt.Fprintf(buf, "%d", t.Time.Day())
+	case 'j':
+		fmt.Fprintf(buf, "%03d", t.Time.YearDay())
+	case 'H':
+		fmt.Fprintf(buf, "%02d", t.Time.Hour())
+	case 'k':
+		fmt.Fprintf(buf, "%d", t.Time.Hour())
+	case 'h', 'I':
+		t := t.Time.Hour()
+		if t == 0 || t == 12 {
+			fmt.Fprintf(buf, "%02d", 12)
+		} else {
+			fmt.Fprintf(buf, "%02d", t%12)
+		}
+	case 'l':
+		fmt.Fprintf(buf, "%d", (t.Time.Hour()%12)+1)
+	case 'i':
+		fmt.Fprintf(buf, "%02d", t.Time.Minute())
+	case 'p':
+		if t.Time.Hour() < 12 {
+			buf.WriteString("AM")
+		} else {
+			buf.WriteString("PM")
+		}
+	case 'r':
+		h := t.Time.Hour()
+		switch {
+		case h == 0:
+			fmt.Fprintf(buf, "%02d:%02d:%02d AM", 12, t.Time.Minute(), t.Time.Second())
+		case h == 12:
+			fmt.Fprintf(buf, "%02d:%02d:%02d PM", 12, t.Time.Minute(), t.Time.Second())
+		case h < 12:
+			fmt.Fprintf(buf, "%02d:%02d:%02d AM", h, t.Time.Minute(), t.Time.Second())
+		default:
+			fmt.Fprintf(buf, "%02d:%02d:%02d PM", h-12, t.Time.Minute(), t.Time.Second())
+		}
+	case 'T':
+		fmt.Fprintf(buf, "%02d:%02d:%02d", t.Time.Hour(), t.Time.Minute(), t.Time.Second())
+	case 'S', 's':
+		fmt.Fprintf(buf, "%02d", t.Time.Second())
+	case 'f':
+		fmt.Fprintf(buf, "%06d", t.Time.Microsecond())
+	case 'U', 'u', 'V', 'v':
+		// TODO: Fix here.
```

Comment:
`TODO: Fix here.` is a special comment I use, so grep can find it easily.
mysql may use Sunday or Monday as the first day of week, `U u V v` controls which.

but go always use Sunday as the first day of week, so not easy to do it now, I leave a TODO.